# RNA-Seq differential expression analysis using Kallisto and DESeq2
Here we describe a pipeline to analyse RNA sequencing data (NovaSeq 6000 PE150). Overall, we need to 1) check the quality of raw data (FASTQC analyisis), 2) filter te raw data (FASTP), and 3) estimate the gene expression of the RNAseq data (KALLISTO). Kallisto is a program for quantifying abundances of transcripts from bulk RNA-Seq data. It is based on the pseudoalignment of reads against the transcriptome of a species for rapid estimation of transcript expression. Of note, this pipeline is described considering that the genome/transcriptome of the species that we are working with is publicly available.

## 0. Download files
The first step is downloading the raw sequencing files into our server of computer. Most sequencing providers will upload the raw files into an FTP server. The files can usually be downloaded using the Unix command "wget":
```
wget https://...
```
Paired-end sequencing generates two output fastq files for each sample. These usually end in "_1.fastq.gz" and "_2.fastq.gz" (but "_R1_001.fastq.gz" and "_R2_001.fastq.gz" are also frequent). The files are usually compressed using gunzip - there is no need to unzip them, most software will work with gunzipped files. The fastq files will usually come with corresponding MD% files to check that the integrity of the downloaded files.


## 1. Quality control with FastQC 
Once we have downloaded the fastq files, it is good practice to check their quality. It can be done using the software [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). The command to run FastQC in a fastq file is:

```
fastqc sampleA_1.fastq.gz
```
Most of the time, it will be more convinient to run FastQC for all samples at once:
```
fastqc *fastq.gz
```
Two files will be generated by FastQC: fastqc.zip and fastqc.html. To visualise them, the easiest way is to open the html files with a web browser. You can find output examples in the webpage of [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/). RNA-Seq data usually present warnings for some of the categories, such as "Per base sequence context" or "Sequence duplication levels". As long as the issues are not extrime, this is normal and nothing to worry about.

This quality control is just to visually inspect the quality of the sequencing data, but it does not alter the fastq files in any way. In fact, this step can even be skipped, although it is worth practice to check the quality of the sequencing output to discards any problems connected to RNA extraction, library preparation or sequencing.


## 2. Filtering using fastp
While not absolutely necessary, removing low quality reads and bases can improve downstream results. [Fastp](https://github.com/OpenGene/fastp) can rapidly filter fastq files, removing low quality reads, contaminating sequence, low complexity reads (repeats), short reads, etc. An example command would be:
```
fastp -i sampleA_1.fastq.gz -I sampleA_2.fastq.gz -o sampleA_filtered_1.fastq.gz -O sampleA_filtered_2.fastq.gz -q 15 -l 30 -h sampleA.html
```
Fastp takes two paired fastq files (options -i and -I) and removes bases with Phred quality below 15 (-q 15) and reads that end up with a lenght of less than 30 bases (-l 30). The raw files remain intact, and two new filtered files are generated (options -o and -O). -h specifies names for the html file with plots showing the read quality before and after filtering  (potential alternative to FastQC).


## 3. Quantification using kallisto
[Kallisto](https://pachterlab.github.io/kallisto/about) is a software for rapidly quantifying abundances of transcripts from RNA-seq data. It is based on a pseudoalignment strategy using the transcriptome of a species as reference. This reference transcriptome needs to be downloaded and indexed before expression can be quantified. Transcriptomes for species with sequenced genomes can be obtained from Ensembl or NCBI (and downloaded using "wget").

To create the index of the transcriptome, the command is:
```
kallisto index -i transcriptome.fa.idx transcriptome.fa
```.


#You have to indicate both files (forward and reverse)/each sample (--in 1 and --in 2), and also the folder (create it in advance) and names of the output files that are still paired (--out1, --out2). --unpaired1, --unpaired2 indicate the output files for the reads that are not 'paired'. -l 35: this specifies that if a read is shorter than 50 basepairs after all filters, it should be removed. -q: quality threshold per base required (default: 15, which means that a Phred quality score of at least 15 is required) -h: specifies name for the html file with plots showing the read quality before and after filtering 
```
fastp --in1 raw_data/E283_1.fq.gz --in2 raw_data/E283_2.fq.gz --out1 filtered/E283_1.filtered.fastq.gz --out2 filtered/E283_2.filtered.fastq.gz --unpaired1 filtered/E283_1.unpaired.fastq.gz --unpaired2 filtered/E283_2.unpaired.fastq.gz -l 35 -q 20 -h filtered/E283.html &> filtered/E283.log
```

#ESTIMATE THE EXPRESSION - KALLISTO
#You will now match the transcriptome with the FILTERED data to obtain the gene expression of your RNAseq. Do this from the folder 'RABBIT'. Kallisto will create the '-o folder' inside 'RABBIT', so don't create it in advance, and you will 'call' your transcriptome with '-i'. The 'transcriptome' folder should be inside 'RABBIT'. -b 100 indicates the maximum lenght of a read (not sure about this?Â¿). Finally you have to indicate the directory of the filtered data (both paired reads)
#You can do that with a compute session if you have few samples. Else, you can queue them with .sh and 'sbatch' (same procedure as for FASTP)
#How to create a .sh file to run kallisto? You have to do a different script for each sample and run them independently. See below an example for ONE SAMPLE.
```
#!/bin/sh
#SBATCH -p thinnodes
#SBATCH -N 1
#SBATCH -n 1
#SBATCH -c 8
#SBATCH -t 2:00:00
#SBATCH --mail-type=begin
#SBATCH --mail-type=end
#SBATCH --mail-user=paularodriguez.villamayor@usc.es

module load cesga/2018 gcccore/6.4.0 kallisto/0.46.1

kallisto quant -i transcriptome/transcriptome_rabbit_index_kallisto.idx -o kallisto_results_E283 -b 100 /mnt/lustre/scratch/home/usc/ge/prv/fetos/filtered/E283_1.filtered.fastq.gz /mnt/lustre/scratch/home/usc/ge/prv/fetos/filtered/E283_2.filtered.fastq.gz
```

#You convert the .txt document into .sh 
#I created the .txt in Windows, then converted into .sh and then transfered to Linux #You need to use the command dos2unix to transfer from windows to linux
```
dos2unix name.sh
```
#Run the file
```
sbatch name.sh #it is important to do this from the folder 'RABBIT' but it will depend on how you organize your folders
```
#Results: Each sample will have its own folder (i.e. kallisto_results_E283) and inside that folder you will find 3 files: abundance.h5, abundance.tsv, and run_info.json





